### ISSUE TITLE
Provide a DbConn fairing for testing

### ISSUE THREAD

**Issue by hschink** [SergioBenitez/Rocket#1628]
First of all, I'd like to say thank you for 🚀! Although I am a Rust rookie, I was able to get the necessary API including automated tests done within a few days. 👍 

## Feature Request

**As** a developer
**I need** a `DbConn` stub
**so** I can run my unit tests without an actual database connection. 

### Why you believe this feature is necessary

At the moment, I need to execute the following statements to test endpoints accessing the database.
```Rust
let rocket = rocket::ignite()
    .attach(DbConn::fairing())
    .mount("/", routes![my::endpoint]);
let client = Client::new(rocket).expect("valid rocket instance");
```
However, using `.attach(DbConn::fairing())` establishs an actual database connection (pool?). Thus, I cannot test my endpoints without a working database connection.

### A convincing use-case for this feature

It is good practice to isolate different parts of the system to facilitate testing.

Expecting a working database connection makes (unit) testing endpoints more complex. Working database connections should only be necessary for integration testing.

### Why this feature can't or shouldn't exist outside of Rocket

An alternative approach is to use an embedded database such as SQLite. Using an embedded database would save me from setting up a database server just for testing. However, that would mean that you have to set up an alternative database just for testing if you do not already use an embedded database.

For the above reason I opt for implementing a `DbConn::stub_fairing()` for testing.

### Final remarks

If you give me some advice, I would volunteer to implement this feature.

---

I hope my request makes sense. Please let me know if I shall provide additional information.

I am looking forward to your feedback.

**Comment by SergioBenitez** [comment]
I don't think this is possible without _massive_ effort. Let me explain.

Let's say we have something like the following route:

```rust
#[get("/<id>")]
async fn read(db: Db, id: i64) -> Option<Json<Post>> {
    sql!("SELECT id, title, text FROM posts WHERE id = ?", id).await.ok().map(Json)
}
```

At runtime, with a real database, the `SELECT` query will be sent off to the database. The database will parse the query, execute it, and return the results. The application finishes this off by marshalling the response as JSON.

Let's say Rocket had some *magic* that let us swap out a real database for a "fake" database, as you propose. In order for the application to work as expected, the "fake" database would have to do everything the "real" database did. In particular, it would have to, well, be the full fledged database! In this example, the "fake" database would need to parse and execute SQL. But this is just _one_ such database. What if the query were in some other language? Or of a completely different nature?

So it seems we're back to square-one wherein we need a real database, or something that acts exactly like the real thing.

The crux of the matter is this: if you need to test how the database reacts, then you need something that emulates the database exactly, or darn close to it. It seems, then, that your problem is not necessarily the availability of the database, as this is unsolvable, but the ease in which it can be made available. For example, as you mention, SQLite lets you easily create in-memory databases. You can then configure the database separately for test profiles, using an in-memory database, which will closely mirror the production profile. Connections will always succeed, and you can test at will.

So what do we do when our database of choice doesn't make it quite as easy as SQLite? I see a few options:
  1) If your queries are database agnostic and standard SQL, use sqlite during testing via `cfg(test)`.
  2) See if there's some kind of in-memory version of the database, perhaps a proxy of some sort.
  3) Bite the bullet and have a test instance available. This is the standard approach.

> Expecting a working database connection makes (unit) testing endpoints more complex. Working database connections should only be necessary for integration testing.

Finally, I want to comment here: you can unit test any part of your Rocket application that doesn't use a database connection by simply calling the route function directly. Almost all of Rocket's types, like `State`, allow themselves to be created directly, facilitating this kind of unit testing.

**Comment by hschink** [comment]
Ok, I missed some important details. Sorry, my fault.

When I wrote the above request, I was implicitly assuming that the actual database interaction is mocked away. Let me give you an actual example from my (not yet pushed to GitHub) code.
```Rust
// skipped included for brevity
#[get("/bikes")]
pub fn get_bikes(db: DbConn) -> Result<Json<Vec<BikeTranslatable>>,diesel::result::Error> {
    let data = bike::get_bikes(&db);

    match data {
        Ok(v) => Ok(Json(v)),
        Err(e) => Err(e),
    }
}

#[cfg(test)]
mod test {

    #[test]
    fn test_get_bikes() {
        crate::database::test::setup();

        bike::get_bikes.mock_safe(|_| MockResult::Return(Ok(vec![
            BikeTranslatable { id: 1,
                bike_id: 1,
                locale: "de-DE".to_string(),
                title: "Test".to_string(),
                description: None,
                url: None
            },
            BikeTranslatable { id: 2,
                bike_id: 1,
                locale: "de-DE".to_string(),
                title: "Test 2".to_string(),
                description: Some("Test description".to_string()),
                url: Some("https://bikes.test.rs/2".to_string())
            },
        ])));

        let rocket = rocket::ignite()
            .attach(DbConn::fairing())
            .mount("/", routes![super::get_bikes]);
        let client = Client::new(rocket).expect("valid rocket instance");

        let mut response = client.get("/bikes").dispatch();
        assert_eq!(response.status(), Status::Ok);
        assert_eq!(response.body_string(), Some("[{\"id\":1,\"bike_id\":1,\"locale\":\"de-DE\",\"title\":\"Test\",\"description\":null,\"url\":null},{\"id\":2,\"bike_id\":1,\"locale\":\"de-DE\",\"title\":\"Test 2\",\"description\":\"Test description\",\"url\":\"https://bikes.test.rs/2\"}]".to_string()));
    }
}
```
I use [mocktopus](https://docs.rs/mocktopus/0.7.11/mocktopus/) here. With `bike::get_bikes.mock_safe(|_| MockResult::Return(//snip`, I am mocking the call to the data access object (DAO). The DAO implements the actual database interaction. You see that with `|_|` I ignore the first parameter, which is `DbConn` in this case. I simply return a vector of expected results.

In short, I expect `DbConn::stub_fairing()` to provide a valid `DbConn` object, so I can call the route for testing without establishing an actual database connection. I do not expect `DbConn::stub_fairing()` to implement the actual database interaction. For the latter, I need to use a mocking framework of my choice.

I hope it makes sense now to have a `DbConn::stub_fairing()` for (unit) testing.

Please let me know if anything is still unclear.

**Comment by SergioBenitez** [comment]
If I'm understanding correctly, you want the following:
  * A database fairing that doesn't try to connect to the database.
  * A database request guard that always succeeds even when there's no connection.

Beyond the second part probably not working the way you expect, since we wouldn't be able to "fake" the underlying connection type, this feels remarkably error prone to me, so I'm immediately inclined to find another solution.

Here's an idea, completely external to Rocket:

```rust
// A mockable DB handle.
#[cfg(test)]
mod db {
    struct Conn;

    #[cfg(test)]
    impl Conn {
        fn fairing() -> AdHoc { /* return a fairing that does nothing */ }
    }

    impl FromRequest for Conn {
        /* infallible implementation */
    }
}

// The real handle.
#[cfg(not(test))]
mod db {
    #[cfg(not(test))]
    #[database("my-db")]
    struct Conn(real::Connection);
}


#[launch]
fn rocket() -> _ {
    rocket::build().attach(db::Conn::fairing());
}
```

This is compile-time checked and forces you to mock all of the underlying connection functionality. It is also entirely external to Rocket.

**Comment by hschink** [comment]
Ok, this looks like a suitable approach. Thanks a lot! :+1:

I just started to implement this and it seems that this will take me some time. I guess we can close this request for now.

When I get a working solution, I will post it here.

**Comment by hschink** [comment]
I looked into @SergioBenitez's [suggestion above](https://github.com/SergioBenitez/Rocket/issues/1628#issuecomment-832078125) and made the following two findings which I'd like to document here.

For implementing your own database handle (called `Conn` in @SergioBenitez's suggestion), it is necessary to create a database connection stub implementing the [Diesel connection trait](https://docs.diesel.rs/master/diesel/connection/trait.Connection.html). However, implementing such a stub is out of scope of Rocket or the application using Rocket and should be done in Diesel itself. At the moment, I cannot assess how much effort is necessary to implement a Diesel connection stub. But it is noteworthy that for Diesel it is recommended to use an actual database connection for testing (see https://github.com/diesel-rs/diesel/issues/1549). Thus, I assume that Diesel does not support this, yet.

Additionally, for integration tests we still need an actual database connection. However, considering we have a database handle stub available, unit _and_ integration tests would both use the stub. At the moment, I do not know if there is a way to conditionally compile the database handle stub only for unit tests and to compile the regular database handle for integration tests without splitting the tests into features.

**tl;dr** Go with @SergioBenitez's [advice above](https://github.com/SergioBenitez/Rocket/issues/1628#issuecomment-831888282):
> 1. If your queries are database agnostic and standard SQL, use sqlite during testing via cfg(test).
> 2. See if there's some kind of in-memory version of the database, perhaps a proxy of some sort.
> 3. Bite the bullet and have a test instance available. This is the standard approach.
